### PR TITLE
Merge main (again)

### PR DESCRIPTION
**Pull request by thurstonsand**
This caused a lot of changes, and a crazy type bound for `IPAInputRow`, as seen here:

```rust
impl<F: Field, MK: BitArray, BK: BitArray> Serializable for IPAInputRow<F, MK, BK>
where
    XorReplicated<BK>: Serializable,
    XorReplicated<MK>: Serializable,
    Replicated<F>: Serializable,
    <XorReplicated<BK> as Serializable>::Size: Add<<Replicated<F> as Serializable>::Size>,
    <Replicated<F> as Serializable>::Size:
        Add<
            <<XorReplicated<BK> as Serializable>::Size as Add<
                <Replicated<F> as Serializable>::Size,
            >>::Output,
        >,
    <XorReplicated<MK> as Serializable>::Size: Add<
        <<Replicated<F> as Serializable>::Size as Add<
            <<XorReplicated<BK> as Serializable>::Size as Add<
                <Replicated<F> as Serializable>::Size,
            >>::Output,
        >>::Output,
    >,
    <<XorReplicated<MK> as Serializable>::Size as Add<
        <<Replicated<F> as Serializable>::Size as Add<
            <<XorReplicated<BK> as Serializable>::Size as Add<
                <Replicated<F> as Serializable>::Size,
            >>::Output,
        >>::Output,
    >>::Output: ArrayLength<u8>,
{
```
and those bounds leaked to other functions -- this had to specified multiple times (as commented in this PR)

### CODE REVIEW

File: src/cli/playbook/ipa.rs [hunk 1]
```diff
@@ -4,37 +4,70 @@ use crate::{
...
+pub async fn semi_honest<F, MK, BK>(
     input: InputSource,
     clients: &[MpcHelperClient; 3],
     query_id: QueryId,
 ) -> [Vec<impl Send + Debug>; 3]
 where
-    F: Field,
-    B: BitArray,
-    Standard: Distribution<F> + Distribution<B>,
+    F: Field + IntoShares<Replicated<F>>,
```
Comment (thurstonsand) [inline src/cli/playbook/ipa.rs:30]: wow this is rough. i need to redefine the trait bounds at the top level usage

File: src/query/executor.rs [hunk 2]
```diff
@@ -98,14 +104,43 @@ async fn execute_test_multiply<F: Field>(
...
 
-async fn execute_ipa<F: Field, B: BitArray>(
+async fn execute_ipa<F: Field, MK: BitArray, BK: BitArray>(
     ctx: SemiHonestContext<'_, F>,
     query_config: IPAQueryConfig,
     mut input: AlignedByteArrStream,
-) -> Vec<AggregateCreditOutputRow<F>> {
+) -> Vec<MCAggregateCreditOutputRow<F, BK>>
+where
+    IPAInputRow<F, MK, BK>: Serializable,
```
Comment (thurstonsand) [inline src/query/executor.rs:113]: here as well

File: src/protocol/ipa/mod.rs [hunk 3]
```diff
@@ -188,44 +255,68 @@ impl<F: Field + Sized> Resharable<F> for IPAModulusConvertedInputRow<F> {
...
     max_breakdown_key: u128,
     num_multi_bits: u32,
-) -> Result<Vec<AggregateCreditOutputRow<F>>, Error> {
-    let mk_shares = input_rows
+) -> Result<Vec<MCAggregateCreditOutputRow<F, BK>>, Error>
+where
+    F: Field,
+    MK: BitArray,
+    BK: BitArray,
+    <F as Serializable>::Size: Add<<F as Serializable>::Size>,
```
Comment (thurstonsand) [inline src/protocol/ipa/mod.rs:269]: needed these as well

File: src/cli/playbook/ipa.rs (hunk 1 above)
Comment (akoshelev) [reply src/cli/playbook/ipa.rs:30]: oh, is it worth doing then? we serialize it into a vec already, so there seem to be no point of trying to make it extremely efficient

File: src/ff/prime_field.rs [hunk 4]
```diff
@@ -221,7 +199,8 @@ mod fp31 {
 }
 
 mod fp32bit {
-    field_impl! { Fp32BitPrime, u32, 4_294_967_291 }
+    use typenum::U5;
+    field_impl! { Fp32BitPrime, u32, 4_294_967_291, U5 }
```
Comment (akoshelev) [inline src/ff/prime_field.rs:203]: Martin fixed this part, it should be `U4`

File: src/query/executor.rs (hunk 2 above)
Comment (akoshelev) [reply src/query/executor.rs:113]: same advice - lets assume IPAInputRow is DST

**Review by akoshelev — APPROVED** [review]

**Comment by akoshelev** [comment]
I am going to merge it as there is no point of iterating on large PRs. Makes it very hard. Lets see if we can make `IpaInputRow` definition a bit easier to swallow in a follow up

**Pull request merged**